tests/test_domain_py.py

```diff
@@ -44,3 +46,99 @@ def test_function_signatures():
+    assert_called(None, u'NestedParentB', u'NestedParentB', u'class')
+    assert_called(None, None, u'NestedParentA.NestedChildA', u'class')
+
+    assert_called('module_a.submodule', 'ModTopLevel', 'mod_child_1', 'meth')
+    assert_called('module_a.submodule', 'ModTopLevel',
+                  'ModTopLevel.mod_child_1', 'meth')
+    assert_called('module_a.submodule', 'ModTopLevel', 'mod_child_2', 'meth')
+    assert_called('module_a.submodule', 'ModTopLevel',
+                  'module_a.submodule.ModTopLevel.mod_child_1', 'meth')
+
+    with pytest.raises(AssertionError):
+        assert_called('module_a.submodule', None, 'ModTopLevel', 'class')
+    with pytest.raises(AssertionError):
+        assert_called('module_b.submodule', None, 'ModTopLevel', 'class')
+    with pytest.raises(AssertionError):
+        assert_called('module_b.submodule', 'ModTopLevel', 'ModNoModule', 'class')
+
+
+@pytest.mark.sphinx(testroot='domain-py')
+def test_domain_py_objects(app, status, warning):
+    app.builder.build_all()
+
+    modules = app.env.domains['py'].data['modules']
+    objects = app.env.domains['py'].data['objects']
+
+    assert 'module_a.submodule' in modules
+    assert 'module_a.submodule' in objects
+    assert 'module_b.submodule' in modules
+    assert 'module_b.submodule' in objects
+
+    assert objects['module_a.submodule.ModTopLevel'] == ('module', 'class')
+    assert objects['module_a.submodule.ModTopLevel.mod_child_1'] == ('module', 'method')
+    assert objects['module_a.submodule.ModTopLevel.mod_child_2'] == ('module', 'method')
+    assert objects['ModNoModule'] == ('module', 'class')
+    assert objects['module_b.submodule.ModTopLevel'] == ('module', 'class')
+
+    assert objects['TopLevel'] == ('roles', 'class')
+    assert objects['top_level'] == ('roles', 'method')
+    assert objects['NestedParentA'] == ('roles', 'class')
+    assert objects['NestedParentA.child_1'] == ('roles', 'method')
+    assert objects['NestedParentA.any_child'] == ('roles', 'method')
+    assert objects['NestedParentA.NestedChildA'] == ('roles', 'class')
+    assert objects['NestedParentA.NestedChildA.subchild_1'] == ('roles', 'method')
+    assert objects['NestedParentA.NestedChildA.subchild_2'] == ('roles', 'method')
+    assert objects['child_2'] == ('roles', 'method')
+    assert objects['NestedParentB'] == ('roles', 'class')
+    assert objects['NestedParentB.child_1'] == ('roles', 'method')
+
+
+@pytest.mark.sphinx(testroot='domain-py')
+def test_domain_py_find_obj(app, status, warning):
+
+    def find_obj(modname, prefix, obj_name, obj_type, searchmode=0):
+        return app.env.domains['py'].find_obj(
+            app.env, modname, prefix, obj_name, obj_type, searchmode)
+
+    app.builder.build_all()
+
+    assert find_obj(None, None, u'NONEXISTANT', u'class') == []
+    assert find_obj(None, None, u'NestedParentA', u'class') == [(
+        u'NestedParentA', (u'roles', u'class'))]
```

Comment:
I feel this is a bit unreadable.
How about this?
```
    assert (find_obj(None, None, u'NestedParentA', u'class') ==
                     [(u'NestedParentA', (u'roles', u'class'))])
```
I like parenthesis are paired (preferably).
If so, we can add easily the case which `find_obj()` returns multiple objects.